### PR TITLE
Support partial generic specialization of closures

### PR DESCRIPTION
**Pull request by swiftix**
Clean-up the code and fix a bug on the way.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 1]
```diff
@@ -130,10 +130,15 @@ bool ReabstractionInfo::prepareAndCheck(ApplySite Apply, SILFunction *Callee,
   bool HasConcreteGenericParams = false;
   bool HasNonArchetypeGenericParams = false;
   HasUnboundGenericParams = false;
-  for (auto DT : CalleeGenericSig->getGenericParams()) {
+  auto *SM = Apply.getModule().getSwiftModule();
+  for (auto GP : CalleeGenericSig->getGenericParams()) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Instead you can call GenericSignature::getSubstitutableParams(), which skips those that are concrete

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 2]
```diff
@@ -382,6 +388,12 @@ ReabstractionInfo::createSubstitutedType(SILFunction *OrigF,
     Lowering::GenericContextScope GenericScope(M.Types,
                                                CanSpecializedGenericSig);
     FnTy = OrigF->getLoweredFunctionType()->substGenericArgs(M, SubstMap);
+    // Some of the added new requirements may not have been taken into account
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Can you add a FIXME here so that I can remember to remove this when I check in my fix?

Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: You have a fix for this?

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 1 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Oh, nice that we have this API now.

**Comment by slavapestov** [comment]
Looks good, thanks!

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Pull request merged**